olive/passes/qnn/context_binary_generator.py

```diff
@@ -0,0 +1,86 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import logging
+import platform
+from pathlib import Path
+from typing import Any, Callable, Dict, Union
+
+from olive.constants import ModelFileFormat
+from olive.hardware import AcceleratorSpec
+from olive.model import ONNXModelHandler, PyTorchModelHandler, QNNModelHandler, TensorFlowModelHandler
+from olive.passes.olive_pass import Pass
+from olive.passes.pass_config import PassConfigParam
+from olive.platform_sdk.qualcomm.runner import QNNSDKRunner
+
+logger = logging.getLogger(__name__)
+
+
+class QNNContextBinaryGenerator(Pass):
+    @staticmethod
+    def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        if platform.system() == "Windows":
+            raise NotImplementedError("QNNContextBinaryGenerator is not supported on Windows.")
+
+        return {
+            "backend": PassConfigParam(
+                type_=str,
+                required=True,
+                description=("Path to a QNN backend .so library to create the context binary."),
+            ),
+            "binary_file": PassConfigParam(
+                type_=str,
+                required=False,
+                description=(
+                    "Name of the binary file to save the context binary to."
+                    " Saved in the same path as --output_dir option with .bin"
+                    " as the binary file extension. If not provided, no backend binary"
+                    " is created."
+                ),
+            ),
+            "extra_args": PassConfigParam(
+                type_=str, default_value=None, description="Extra arguments to qnn-context-binary-generator"
+            ),
+        }
+
+    @staticmethod
+    def _validators() -> Dict[str, Callable[..., Any]]:
+        pass
+
+    def _run_for_config(
+        self,
+        model: Union[TensorFlowModelHandler, PyTorchModelHandler, ONNXModelHandler],
+        data_root: str,
+        config: Dict[str, Any],
+        output_model_path: str,
+    ) -> QNNModelHandler:
+        main_cmd = "qnn-context-binary-generator"
+        runner = QNNSDKRunner(dev=True)
```

Comment:
what's meaning dev=True for all cases?

Reply:
`dev=True` means run the cmd locally. Run QNN optimization has following steps:
1. convert and build model in x64 windows/linux (dev = True
2. set the model to target device(arm-windows or android) (dev=False

Reply:
If it is used for run locally or remotely, suggest to change the name to local_run or something else. 

In my mind, dev=True means run the optimization using the local branch Olive bits.

Reply:
Updated with `optional_local_run` where the True value is used to tell current runner can run cmd locally.
The value is set to False only if the `cmd` have to be run remotely but cannot in local env.